homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,180 @@
+
+
+def washer_state(washer: WasherDryer) -> str | None:
+    """Determine correct states for a washer."""
+
+    machine_state = washer.get_machine_state()
+    machine_cycle = None
+    for count, func in zip(range(len(CYCLE_FUNC)), CYCLE_FUNC):
+        if func(washer):
+            machine_cycle = WASHER_STATE.get(count)
+            break
+
+    if washer.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        machine_cycle = "Door open"
+        return machine_cycle
+
+    if machine_state == MachineState.RunningMainCycle and machine_cycle:
+        return machine_cycle
+
+    return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
+
+
+@dataclass
+class WhirlpoolSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable
+
+
+@dataclass
+class WhirlpoolSensorEntityDescription(
+    SensorEntityDescription, WhirlpoolSensorEntityDescriptionMixin
+):
+    """Describes Whirlpool Washer sensor entity."""
+
+
+SENSORS: tuple[WhirlpoolSensorEntityDescription, ...] = (
+    WhirlpoolSensorEntityDescription(
+        key="state",
+        name="state",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=washer_state,
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="timeremaining",
+        name="time remaining",
+        device_class=SensorDeviceClass.DURATION,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement="s",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=lambda WasherDryer: WasherDryer.get_attribute(
+            "Cavity_TimeStatusEstTimeRemaining"
+        ),
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="DispenseLevel",
+        name="Dispense Level",
+        device_class=SensorDeviceClass.VOLUME,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=lambda WasherDryer: TANK_FILL[
+            WasherDryer.get_attribute("WashCavity_OpStatusBulkDispense1Level")
+        ],
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Config flow entry for Whrilpool Laundry."""
+    whirlpool_data: WhirlpoolData = hass.data[DOMAIN][config_entry.entry_id]
+    for appliance in whirlpool_data.appliances_manager.washer_dryers:
+        _wd = WasherDryer(
+            whirlpool_data.backend_selector,
+            whirlpool_data.auth,
+            appliance["SAID"],
+        )
+        await _wd.connect()
+        entities = [
+            WasherDryerClass(
+                appliance["SAID"],
+                appliance["NAME"],
+                description,
+                _wd,
+            )
+            for description in SENSORS
+        ]
+
+        async_add_entities(entities)
```

Comment:
I think `async_add_entities` should be called once per platform. Also, it is a best practice to start all the tasks in parallel then do an `asyncio.gather` at the end to wait for them all to finish together which may help do it all at once.